cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,564 @@
+        int totalsize = size_x * size_y * size_z;
+
+        auto tile_inserted_label_mirror = Kokkos::create_mirror_view(
+            Kokkos::HostSpace(), _tile_inserted_label );
+        for ( int i = 0; i < totalsize; i++ )
+        {
+            tile_inserted_label_mirror( i ) = 0;
+        }
+        Kokkos::deep_copy( _tile_inserted_label, tile_inserted_label_mirror );
+
+        _rehash_coeff[0] = rehash_factor;
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType valid_block_num() const { return _tile_table_info( 0 ); }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    void insert( int tile_i, int tile_j, int tile_k ) const
+    {
+        // reallocation after insert
+        KeyType tileKey = ijk_to_key( tile_i, tile_j, tile_k );
+        ValueType tileNo = (ValueType)0;
+        if ( Kokkos::atomic_fetch_or(
+                 &( _tile_inserted_label( tile_i * _block_size[1] *
+                                              _block_size[2] +
+                                          tile_j * _block_size[2] + tile_k ) ),
+                 1 ) == 0 )
+        {
+            // current impl: use inserting sequence as id
+            // rethink: sort the id .
+            tileNo = Kokkos::atomic_fetch_add( &( _tile_table_info( 0 ) ),
+                                               (ValueType)1 );
+            _tile_table.insert( tileKey, tileNo );
+        }
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    void sortTile()
+    {
+        // pass
+    }
+
+    // query
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType find( int tile_i, int tile_j, int tile_k ) const
+    {
+        return _tile_table.value_at(
+            _tile_table.find( ijk_to_key( tile_i, tile_j, tile_k ) ) );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    KeyType ijk_to_key( int tile_i, int tile_j, int tile_k ) const
+    {
+        return _op_ijk2key( tile_i, tile_j, tile_k );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    void key_to_ijk( KeyType &key, int &tile_i, int &tile_j, int &tile_k ) const
+    {
+        return _op_key2ijk( key, tile_i, tile_j, tile_k );
+    }
+
+  private:
+    //! [0] current valid table size, [1] pre-allocated size
+    Kokkos::View<int[2], ExecutionSpace> _tile_table_info;
+    Kokkos::View<int *, ExecutionSpace> _tile_inserted_label;
+    Kokkos::Array<int, 3> _block_size;
+    //! default factor, rehash by which when need more capacity of the hash
+    //! table
+    Kokkos::Array<float, 1> _rehash_coeff;
+    //! hash table (tile IJK <=> tile No)
+    Kokkos::UnorderedMap<KeyType, ValueType, ExecutionSpace> _tile_table;
+    //! Op: tile IJK <=> tile No
+    TileID2HashKey<KeyType, HashType> _op_ijk2key;
+    HashKey2TileID<KeyType, HashType> _op_key2ijk;
+};
+
+//---------------------------------------------------------------------------//
+template <int CBits, int CNumPerDim, int CNumPerTile>
+class TileIndexSpace
+{
+  public:
+    //! Number of cells inside each tile
+    static constexpr int CellBitsPerTileDim = CBits;
```

Comment:
We consider static class members like this to be variables so I would do `cell_bits_per_tile_dim` for this and other objects like it. `CBits` is OK because it is a template parameter and we do upper-case camel case for the template parameters.

Reply:
Ours is a little different. Admittedly it is some combination I developed over the years. In hindsight maybe it would be better to use something more well-known but we will stick with this for now.

Reply:
So is there any doc to follow the convention? I can make a changed version based on that so you don't need to comment line by line here.

Reply:
I can add a style guide to the contributing doc if that's the best place

Reply:
Thank you very much for providing this option. I can first follow other files/codes as style guides. I will post here if there are still problems!